### ISSUE TITLE
PKCS8::load_key is affected by a compiler bug in clang 8

### ISSUE THREAD

**Issue by reneme** [randombit/botan#2255]
## TL;DR

This is an informational ticket in the hope of helping other people that might stumble over the issue. A [bug in clang 8](https://bugs.llvm.org/show_bug.cgi?id=41810) can cause programs using Botan to crash when calling `PKCS8::load_key` with an incorrect password of a certain length.

Feel free to close this right away.

## Symptoms

We recently received a bug report claiming that our application crashes for long (>800 characters) user-provided passwords. We quickly realized that crashes occur for (wrong) passwords that are longer than only 23 (!) characters.

This was reproducible and the stack trace looked roughly like that:

```
spike(44594,0x10081d5c0) malloc: *** error for object 0x100a07340: pointer being freed was not allocated
spike(44594,0x10081d5c0) malloc: *** set a breakpoint in malloc_error_break to debug

* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
  * frame #0: 0x00007fff720962c2 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff72151bf1 libsystem_pthread.dylib`pthread_kill + 284
    frame #2: 0x00007fff720006a6 libsystem_c.dylib`abort + 127
    frame #3: 0x00007fff7210f077 libsystem_malloc.dylib`malloc_vreport + 545
    frame #4: 0x00007fff7210ee38 libsystem_malloc.dylib`malloc_report + 151
    frame #5: 0x00000001001b11fc spike`Botan::PKCS8::load_key(Botan::DataSource&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) [inlined] std::__1::_DeallocateCaller::__do_call(__ptr=<unavailable>) at new:320:12 [opt]
    frame #6: 0x00000001001b11f7 spike`Botan::PKCS8::load_key(Botan::DataSource&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) [inlined] std::__1::_DeallocateCaller::__do_deallocate_handle_size(__ptr=<unavailable>) at new:278 [opt]
    frame #7: 0x00000001001b11f7 spike`Botan::PKCS8::load_key(Botan::DataSource&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) [inlined] std::__1::_DeallocateCaller::__do_deallocate_handle_size_align(__ptr=<unavailable>, __align=1) at new:248 [opt]
    frame #8: 0x00000001001b11f7 spike`Botan::PKCS8::load_key(Botan::DataSource&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) [inlined] std::__1::__libcpp_deallocate(__ptr=<unavailable>, __align=1) at new:326 [opt]
    frame #9: 0x00000001001b11f7 spike`Botan::PKCS8::load_key(Botan::DataSource&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) [inlined] std::__1::allocator<char>::deallocate(__p=<unavailable>) at memory:1817 [opt]
    frame #10: 0x00000001001b11f7 spike`Botan::PKCS8::load_key(Botan::DataSource&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) [inlined] std::__1::allocator_traits<std::__1::allocator<char> >::deallocate(__p=<unavailable>) at memory:1555 [opt]
    frame #11: 0x00000001001b11f7 spike`Botan::PKCS8::load_key(Botan::DataSource&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) [inlined] std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::~basic_string(this="") at string:2139 [opt]
    frame #12: 0x00000001001b11ed spike`Botan::PKCS8::load_key(Botan::DataSource&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) [inlined] std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::~basic_string(this="") at string:2134 [opt]
    frame #13: 0x00000001001b11ed spike`Botan::PKCS8::load_key(source=0x00007ffeefbff5d0, pass=<unavailable>) at botan_all.cpp:59728 [opt]
    frame #14: 0x00000001000767a4 spike`Botan::PKCS8::load_key(source=<unavailable>, rng=<unavailable>, pass=<unavailable>) at botan_all.cpp:59771:11 [opt]
    frame #15: 0x00000001003926cc spike`main at spike.cc:27:20 [opt]
    frame #16: 0x00007fff71f5b3d5 libdyld.dylib`start + 1
```

Some sort of double-free inside a `~basic_string<>` destructor invocation was to blame. The culprit destructor is called in the function [`Botan::PKCS8::load_key`](https://github.com/randombit/botan/blob/2.13.0/src/lib/pubkey/pkcs8.cpp#L457) after some exception is thrown deeper down in the call stack. More specifically, a `Botan::Decoding_Error` reporting a broken padding due to the wrong password.

## Investigation

Passwords that are longer than 23 characters kick off the ["small buffer optimization"](https://akrzemi1.wordpress.com/2014/04/14/common-optimizations/) implemented in `std::string` and result in a heap allocation (on 64bit x86 at least). I.e. any string longer than 23 characters caused our program to crash with the above malloc panic when passed in as the password to `Botan::PKCS8::load_key()`.

This is a minimal example that reproduces the crash on 64bit x86 when compiled with clang 8:
```C++
int main()
{
    auto pem = getPrivateKey(); // some PKCS #8 encoded and ENCRYPTED private key
    Botan::System_RNG rng;

    // too long password
    std::string pw = "abcdabcdabcdabcdabcdabcd";

    Botan::DataSource_Memory ds(reinterpret_cast<const uint8_t*>(pem.data()),
                                pem.size());

    try {
        auto key = Botan::PKCS8::load_key(ds, rng, pw);
    } catch (const std::exception &ex) {
        std::cerr << ex.what() << std::endl;
        return 1;
    }

    return 0;
}
```

## Root Cause

The root cause lies in the overloading of the various `Botan::PKCS8::load_key()` implementations. Specifically this one:

```C++
Private_Key* load_key(const std::string& fsname,
                      RandomNumberGenerator& rng,
                      const std::string& pass)
   {
   BOTAN_UNUSED(rng);
   DataSource_Stream in(fsname);
   return PKCS8::load_key(in, [pass]() { return pass; }).release();
   }
```

The variable `pass` is captured by value into a lambda and passed on to a different overload of `Botan::PKCS8::load_key()`. Further up in the call stack, the above-mentioned `Botan::Decoding_Error` is thrown, causing the stack unwinding to eventually destruct the copied `pass` object in the lambda.

Unfortunately, [clang 8 contains a bug](https://bugs.llvm.org/show_bug.cgi?id=41810) that calls the `~basic_string<>` d'tor one too many times in this case. If the password is longer than 23 characters, the `std::string` contains a pointer to a heap-allocated buffer which is then double-freed. The program crashes.

This is NOT a problem in Botan! The issue is [reproducible independently of Botan](https://godbolt.org/z/wvJLT4) (see the program's output in Compiler Explorer).

## Solution

Both clang 7 and clang 9 are not affected by this issue.

## Acknowledgements

Thanks to @Tymolc and @xqp for spending the whole day crowd-debugging with me.

**Comment by randombit** [comment]
Ugh that must have been fun to debug. Interesting. Based on the backtrace you were seeing this on macOS - does this affect some version of Apple's Clang as well?

Based on https://repology.org/project/clang/versions quite a few distros ship with Clang 8 and especially as (per the LLVM bug) no fixed version of Clang 8 is expected *and* because this seems to involve a double free with externally provided input (!) it would be nice to fix this if possible, at least for this one specific case. A quick grep didn't turn up any other problematic looking lambdas but I didn't spend that much time on it. Can we use `std::bind` here instead?

There should be some Travis configuration which allows us to reproduce this. So

- [ ] A test case, should be easy
- [ ] Confirmed reproduction in CI
- [ ] Change to std::bind (?? assuming this fixes it)
- [ ] A comment explaining why std::bind is being used instead of a lambda

**Comment by reneme** [comment]
This happened on macOS, that's correct. We're using a custom compiler over Apple's Clang because we want to use C++17 in our application on older macOS versions as well. Anyway, at least with the compiler shipping with latest Xcode 11 on macOS 10.14 the problem is not reproducible.

As an alternative of using `std::bind` we could change the basic implementation of `load_key` to take a plain `std::string const&` as password. Then we could make the overloads with the `get_pass` callbacks invoke it right away before up-calling into the basic implementation. That would avoid the lambda altogether.

**Comment by randombit** [comment]
Alternative you suggest would work but changes semantics when loading an unecrypted private key - currently the callback is not invoked in that case (this matters if eg the callback is opening a UI element). It's quite possible that literally nobody is depending on that behavior, but as a library it is hard to be sure.

These interfaces are really old, and could use a complete redesign at some point, but in the short term I don't want to do anything that changes behavior.

It looks like Ubuntu 18.04 includes Clang 8 as an optional package so we can probably repro it that way.

**Comment by randombit** [comment]
Fixed with #2277 thank you @hrantzsch @reneme
